File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxi(TreeNode* root, int& sum)
    {
        if(root == NULL)
           return 0;
        int lefty = max(0,maxi(root->left,sum));
        int righty = max(0,maxi(root->right,sum));
        sum = max(sum,lefty+righty+root->val);
        return root->val + max(lefty,righty);   
    }
    int maxPathSum(TreeNode* root) {
        int sum = INT_MIN;
        maxi(root,sum);
        return sum ;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def helper(root):
            if not root :
                return 0,0
            if not root.left and not root.right :
                return root.val  , root.val  
            left_ans , lefty = helper(root.left)
            right_ans , righty = helper(root.right)
            lefty = max(lefty, 0)
            righty = max(righty ,0)
            maxi = max(lefty + root.val+righty , left_ans, right_ans)
            return maxi , max(lefty ,righty)+root.val    

        ans , _ = helper(root)
        return ans
             
        